File: screener/insiders.py
"""Detect promoter/insider holding increases.

TradingView's screener API does not expose promoter/insider holding fields, so
we scan a liquid universe with TradingView, then enrich each ticker from two
complementary sources:

* yfinance ``Ticker.insider_purchases`` — 6-month aggregate of insider buy/sell
  transactions. Available for both US tickers and ``.NS`` Indian listings.
  Positive net shares ⇒ insiders bought more than they sold.

* openscreener (screener.in) — quarterly shareholding pattern (``promoters``,
  ``fiis``, ``diis``). Indian-only. We compute the latest-quarter delta in
  promoter % vs. the previous quarter; positive ⇒ promoter holding increased.

For India we use openscreener as the primary signal (this is the canonical
source for "promoter holding") and yfinance as a secondary cross-check.
For US the yfinance feed is the only signal.
"""

from __future__ import annotations

import json
import logging
import os
import urllib.parse
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Optional

import pandas as pd
import yfinance as yf

from screener.providers import CachedProvider, ProviderSpec
from screener.resilience import call_with_resilience


logger = logging.getLogger(__name__)
_INDIA_SUFFIXES = (".NS", ".BO")
_SCREENER_URL = "https://www.screener.in/company/{symbol}/"
_SCREENER_HEADERS = {"User-Agent": "Mozilla/5.0 (compatible; screener-cli/1.0)"}

_FMP_INSIDER_URL = "https://financialmodelingprep.com/api/v4/insider-trading"
# FMP only exposes Form 3/4/5 (SEC) data, which covers US listings. Indian
# tickers stay on the screener.in promoter feed.
_FMP_WINDOW_DAYS = 182
_FMP_MAX_PAGES = 10

# Provider seams (cache + resilience). Default TTLs match the legacy
# hand-wired call sites; ``cache_ttl`` is overridden per-call to honour CLI
# flags. The yfinance feed shares the "yfinance" breaker, FMP the "fmp"
# breaker, and the screener.in scrape the "screener-in" breaker.
_YF_INSIDER_PROVIDER = CachedProvider(
    ProviderSpec(provider="yfinance", namespace="yfinance_insiders", ttl_seconds=86400)
)
_FMP_INSIDER_PROVIDER = CachedProvider(
    ProviderSpec(provider="fmp", namespace="fmp_insiders", ttl_seconds=86400)
)
_OPENSCREENER_PROVIDER = CachedProvider(
    ProviderSpec(
        provider="screener-in",
        namespace="openscreener_promoters",
        ttl_seconds=7 * 86400,
    )
)


def _tv_to_yf(ticker: str, market: str) -> str:
    symbol = ticker.split(":", 1)[1] if ":" in ticker else ticker
    if market == "india" and not symbol.endswith(_INDIA_SUFFIXES):
        return f"{symbol}.NS"
    return symbol


# ── yfinance insider purchases ─────────────────────────────────────────────


def _row_value(df: pd.DataFrame, label: str, column: str) -> Optional[float]:
    if df is None or df.empty:
        return None
    if "Insider Purchases Last 6m" not in df.columns:
        logger.debug(
            "yfinance insider purchases payload missing label column; columns=%s",
            list(df.columns),
        )
        return None
    match = df[df["Insider Purchases Last 6m"] == label]
    if match.empty:
        return None
    val = match.iloc[0].get(column)
    if val is None or pd.isna(val):
        return None
    try:
        return float(val)
    except (TypeError, ValueError):
        return None


def _fetch_yf_one(
    name: str,
    yf_symbol: str,
    *,
    cache_ttl: float | None,
    refresh: bool,
) -> Optional[dict]:
    def _fetch() -> Optional[dict]:
        purchases = yf.Ticker(yf_symbol).insider_purchases
        if purchases is None or purchases.empty:
            return None

        return {
            "name": name,
            "yf_symbol": yf_symbol,
            "yf_net_shares_6m": _row_value(
                purchases, "Net Shares Purchased (Sold)", "Shares"
            ),
            "yf_net_pct_6m": _row_value(
                purchases, "% Net Shares Purchased (Sold)", "Shares"
            ),
            "yf_total_held": _row_value(
                purchases, "Total Insider Shares Held", "Shares"
            ),
            "yf_buy_trans_6m": _row_value(purchases, "Purchases", "Trans"),
            "yf_sell_trans_6m": _row_value(purchases, "Sales", "Trans"),
        }

    return _YF_INSIDER_PROVIDER.fetch(
        ("insider_purchases", name, yf_symbol),
        _fetch,
        refresh=refresh,
        fallback=None,
        ttl_seconds=cache_ttl,
        operation=f"insider purchases {yf_symbol}",
    )


def fetch_yfinance_insiders(
    universe: pd.DataFrame,
    market: str,
    max_workers: int = 12,
    cache_ttl: float | None = 86400,
    refresh: bool = False,
) -> pd.DataFrame:
    if universe.empty:
        return pd.DataFrame()

    jobs = [
        (str(row["name"]), _tv_to_yf(str(row["ticker"]), market))
        for _, row in universe.iterrows()
    ]
    rows: list[dict] = []
    with ThreadPoolExecutor(max_workers=max_workers) as pool:
        futures = [
            pool.submit(_fetch_yf_one, n, s, cache_ttl=cache_ttl, refresh=refresh)
            for n, s in jobs
        ]
        for fut in as_completed(futures):
            r = fut.result()
            if r is not None:
                rows.append(r)
    return pd.DataFrame(rows)


# ── FMP insider trading (SEC Form 4) ───────────────────────────────────────


def _fmp_api_key() -> Optional[str]:
    """Resolve FMP_API_KEY, loading the project .env once like the backtester."""
    key = os.environ.get("FMP_API_KEY")
    if key:
        return key
    try:
        from screener.backtester.data import load_env_file
    except Exception:
        return None
    load_env_file()
    return os.environ.get("FMP_API_KEY")


def _aggregate_fmp_transactions(
    transactions: list[dict], window_days: int = _FMP_WINDOW_DAYS
) -> Optional[dict]:
    """Aggregate FMP insider rows into 6-month net buy/sell share counts.

    On SEC Form 4 the ``acquistionOrDisposition`` flag (FMP's spelling) only
    says whether shares were acquired (``A``) or disposed (``D``). An ``A`` row
    covers open-market purchases *and* grants, awards, option-exercises and
    gifts, so it cannot stand alone as a "buy" signal. We therefore key off
    the more specific ``transactionType`` SEC code:

    * a genuine **buy** is ``transactionType`` starting with ``P-`` (Purchase)
      *and* ``acquistionOrDisposition == "A"``;
    * a genuine **sell** is ``transactionType`` starting with ``S-`` (Sale)
      *and* ``acquistionOrDisposition == "D"``.

    Every other code (``A-Award``, ``G-Gift``, ``M-Exempt``,
    ``F-Payment of Exercise`` …) is skipped. Net shares > 0 ⇒ insiders bought
    more than they sold over the window. ``transactionType`` may be absent or
    ``None``; it is coerced to ``str`` (default ``""``) so a missing code just
    skips the row instead of raising.
    """
    if not transactions:
        return None
    cutoff = pd.Timestamp.now().normalize() - pd.Timedelta(days=window_days)
    bought = sold = 0.0
    buy_trans = sell_trans = 0
    for txn in transactions:
        date_raw = txn.get("transactionDate") or txn.get("filingDate")
        ts = pd.to_datetime(date_raw, errors="coerce")
        if pd.isna(ts) or ts < cutoff:
            continue
        try:
            shares = float(txn.get("securitiesTransacted") or 0.0)
        except (TypeError, ValueError):
            continue
        disposition = txn.get("acquistionOrDisposition")
        txn_type = str(txn.get("transactionType") or "").upper()
        if txn_type.startswith("P-") and disposition == "A":
            bought += shares
            buy_trans += 1
        elif txn_type.startswith("S-") and disposition == "D":
            sold += shares
            sell_trans += 1
    if buy_trans == 0 and sell_trans == 0:
        return None
    return {
        "fmp_net_shares_6m": bought - sold,
        "fmp_buy_shares_6m": bought,
        "fmp_sell_shares_6m": sold,
        "fmp_buy_trans_6m": buy_trans,
        "fmp_sell_trans_6m": sell_trans,
    }


def _fetch_fmp_insider_one(
    name: str,
    symbol: str,
    *,
    api_key: str,
    cache_ttl: float | None,
    refresh: bool,
) -> Optional[dict]:
    def _fetch() -> Optional[dict]:
        cutoff = pd.Timestamp.now().normalize() - pd.Timedelta(days=_FMP_WINDOW_DAYS)
        truncated = False

        def _request_page(page: int) -> Optional[list]:
            query = urllib.parse.urlencode(
                {"symbol": symbol, "page": page, "apikey": api_key}
            )
            req = urllib.request.Request(
                f"{_FMP_INSIDER_URL}?{query}", headers=_SCREENER_HEADERS
            )
            with urllib.request.urlopen(req, timeout=20) as resp:
                payload = json.loads(resp.read().decode("utf-8", "ignore"))
            return payload if isinstance(payload, list) else None

        # FMP paginates insider rows (newest first). Walk pages until one
        # is empty/non-list, the oldest row on a page predates the 182-day
        # window, or we hit a safety cap (no unbounded loop on bad data).
        collected: list[dict] = []
        expected_page_size: int | None = None
        for page in range(_FMP_MAX_PAGES):
            rows = _request_page(page)
            if not rows:
                break
            if expected_page_size is None:
                expected_page_size = len(rows)
            collected.extend(rows)
            oldest_raw = rows[-1].get("transactionDate") or rows[-1].get("filingDate")
            oldest = pd.to_datetime(oldest_raw, errors="coerce")
            if not pd.isna(oldest) and oldest < cutoff:
                break
            if (
                page == _FMP_MAX_PAGES - 1
                and expected_page_size is not None
                and len(rows) >= expected_page_size
                and not pd.isna(oldest)
                and oldest >= cutoff
            ):
                truncated = True
                logger.warning(
                    "FMP insider trading for %s may be truncated at %d pages",
                    symbol,
                    _FMP_MAX_PAGES,
                )

        transactions = collected or None
        if not transactions:
            return None
        agg = _aggregate_fmp_transactions(transactions)
        if agg is None:
            return None
        # Surface page-cap truncation to callers (not just a log line): the
        # 6m totals may be incomplete when the history was cut at the cap.
        return {"name": name, "fmp_symbol": symbol, "fmp_truncated": truncated, **agg}

    return _FMP_INSIDER_PROVIDER.fetch(
        ("insider_trading", name, symbol),
        _fetch,
        refresh=refresh,
        fallback=None,
        ttl_seconds=cache_ttl,
        operation=f"insider trading {symbol}",
    )


def fetch_fmp_insiders(
    universe: pd.DataFrame,
    market: str,
    max_workers: int = 12,
    cache_ttl: float | None = 86400,
    refresh: bool = False,
) -> pd.DataFrame:
    """Fetch 6-month net insider buying from FMP for each US ticker.

    Returns an empty frame when no ``FMP_API_KEY`` is configured, so callers
    can fall back to the yfinance feed transparently.
    """
    if universe.empty:
        return pd.DataFrame()
    api_key = _fmp_api_key()
    if not api_key:
        return pd.DataFrame()

    jobs = [
        (str(row["name"]), _tv_to_yf(str(row["ticker"]), market))
        for _, row in universe.iterrows()
    ]
    rows: list[dict] = []
    with ThreadPoolExecutor(max_workers=max_workers) as pool:
        futures = [
            pool.submit(
                _fetch_fmp_insider_one,
                n,
                s,
                api_key=api_key,
                cache_ttl=cache_ttl,
                refresh=refresh,
            )
            for n, s in jobs
        ]
        for fut in as_completed(futures):
            r = fut.result()
            if r is not None:
                rows.append(r)
    return pd.DataFrame(rows)


# ── openscreener (screener.in) quarterly promoter % ────────────────────────


class _HttpScraper:
    """Minimal scraper compatible with openscreener's Stock(scraper=...) interface.

    Skips Playwright (the bundled scraper hangs on screener.in's networkidle
    state because of long-tail analytics requests). screener.in returns the
    full shareholding section in the initial HTML response, so a plain HTTP
    GET is sufficient.
    """

    base_url = _SCREENER_URL
    consolidated = False

    def fetch_page(self, symbol: str) -> str:
        def _fetch() -> str:
            req = urllib.request.Request(
                self.base_url.format(symbol=symbol.upper()), headers=_SCREENER_HEADERS
            )
            with urllib.request.urlopen(req, timeout=20) as resp:
                return resp.read().decode("utf-8", "ignore")

        return call_with_resilience(
            "screener-in",
            f"company page {symbol}",
            _fetch,
            fallback="",
        )

    def fetch_pages(self, symbols):
        return {s.upper(): self.fetch_page(s) for s in symbols}


def _fetch_openscreener_one(
    name: str,
    *,
    cache_ttl: float | None,
    refresh: bool,
) -> Optional[dict]:
    def _fetch() -> Optional[dict]:
        try:
            from openscreener import Stock
        except ImportError:
            return None
        rows = Stock(name, scraper=_HttpScraper()).shareholding_quarterly()
        if not rows:
            return None
        if len(rows) < 2:
            return None

        latest, prev = rows[-1], rows[-2]
        p_latest = latest.get("promoters")
        p_prev = prev.get("promoters")
        if p_latest is None or p_prev is None:
            return None
        try:
            change = float(p_latest) - float(p_prev)
        except (TypeError, ValueError):
            return None

        return {
            "name": name,
            "promoter_pct_latest": float(p_latest),
            "promoter_pct_prev": float(p_prev),
            "promoter_change": change,
            "latest_quarter": latest.get("date"),
            "fii_pct_latest": latest.get("fiis"),
            "dii_pct_latest": latest.get("diis"),
        }

    return _OPENSCREENER_PROVIDER.fetch(
        ("shareholding_quarterly", name),
        _fetch,
        refresh=refresh,
        fallback=None,
        ttl_seconds=cache_ttl,
        operation=f"shareholding {name}",
    )


def fetch_openscreener_promoters(
    universe: pd.DataFrame,
    max_workers: int = 6,
    cache_ttl: float | None = 7 * 86400,
    refresh: bool = False,
) -> pd.DataFrame:
    """Fetch quarterly promoter % from screener.in for each Indian ticker."""
    if universe.empty:
        return pd.DataFrame()

    names = universe["name"].astype(str).tolist()
    rows: list[dict] = []
    with ThreadPoolExecutor(max_workers=max_workers) as pool:
        futures = [
            pool.submit(
                _fetch_openscreener_one, n, cache_ttl=cache_ttl, refresh=refresh
            )
            for n in names
        ]
        for fut in as_completed(futures):
            r = fut.result()
            if r is not None:
                rows.append(r)
    return pd.DataFrame(rows)


# ── filtering ──────────────────────────────────────────────────────────────


def filter_promoter_increased(
    insiders: pd.DataFrame,
    market: str,
    min_promoter_change_pct: float = 0.0,
    min_yf_net_pct: Optional[float] = None,
    require_both: bool = False,
) -> pd.DataFrame:
    """Keep tickers where holdings increased.

    For India: require ``promoter_change >= min_promoter_change_pct`` from
    openscreener. If ``require_both`` is set, also require positive yfinance
    net buys.

    For US: require ``yf_net_shares_6m > 0`` (and optional ``min_yf_net_pct``).
    """
    if insiders.empty:
        return insiders

    if market == "india":
        change = pd.to_numeric(insiders.get("promoter_change"), errors="coerce")
        mask = change > min_promoter_change_pct
        if require_both:
            yf_net = pd.to_numeric(insiders.get("yf_net_shares_6m"), errors="coerce")
            mask = mask & (yf_net > 0)
    else:
        # US: FMP (SEC Form 4) is the primary signal when available; fall back
        # to the yfinance feed per-row when FMP has no data for a ticker.
        yf_net = pd.to_numeric(insiders.get("yf_net_shares_6m"), errors="coerce")
        if "fmp_net_shares_6m" in insiders.columns:
            fmp_net = pd.to_numeric(insiders.get("fmp_net_shares_6m"), errors="coerce")
            net = fmp_net.where(fmp_net.notna() & (fmp_net != 0.0), yf_net)
        else:
            net = yf_net
        mask = net > 0
        if min_yf_net_pct is not None:
            pct = pd.to_numeric(insiders.get("yf_net_pct_6m"), errors="coerce")
            mask = mask & (pct >= min_yf_net_pct)

    return insiders[mask.fillna(False)].copy()

File: screener-rs/src/screeners/insiders.rs
use crate::providers::screener_in::ScreenerInClient;
use crate::screeners::models::ScreenRow;
use anyhow::Context;
use chrono::{Duration, NaiveDate, Utc};
use rayon::prelude::*;
use reqwest::blocking::Client;
use serde_json::Value;

const FMP_WINDOW_DAYS: i64 = 182;
const FMP_MAX_PAGES: usize = 10;

#[derive(Debug, Clone)]
pub struct PromoterBuyRequest {
    pub market: String,
    pub limit: usize,
    pub min_change_pct: f64,
    pub min_yf_net_pct: Option<f64>,
    pub require_both: bool,
}

#[derive(Debug, Clone, Default)]
struct FmpAggregate {
    net_shares_6m: f64,
    buy_shares_6m: f64,
    sell_shares_6m: f64,
    buy_trans_6m: usize,
    sell_trans_6m: usize,
}

pub fn screen_promoter_buys(
    universe: &[ScreenRow],
    request: &PromoterBuyRequest,
) -> anyhow::Result<Vec<ScreenRow>> {
    match request.market.as_str() {
        "india" => screen_india_promoters(universe, request),
        "us" => screen_us_insiders(universe, request),
        other => anyhow::bail!("unknown promoter-buys market {other:?}"),
    }
}

fn screen_india_promoters(
    universe: &[ScreenRow],
    request: &PromoterBuyRequest,
) -> anyhow::Result<Vec<ScreenRow>> {
    if request.require_both {
        anyhow::bail!(
            "Rust promoter-buys supports screener.in promoter deltas; --require-both needs Yahoo insider-purchases, which is not exposed by the Rust provider yet"
        );
    }
    if request.min_yf_net_pct.is_some() {
        anyhow::bail!(
            "--min-yf-net-pct is US/yfinance-only and is not supported by Rust promoter-buys"
        );
    }
    let client = ScreenerInClient::new()?;
    let mut out = Vec::new();
    let bases = universe
        .iter()
        .filter_map(|base| {
            base.text("name")
                .or_else(|| base.text("ticker"))
                .map(|name| (base, india_symbol(&name)))
        })
        .collect::<Vec<_>>();
    let symbols = bases
        .iter()
        .map(|(_, symbol)| symbol.clone())
        .collect::<Vec<_>>();
    let snapshots = client.promoter_snapshots(&symbols);
    for ((base, symbol), (_, snapshot)) in bases.into_iter().zip(snapshots) {
        let Some(snapshot) = snapshot else {
            continue;
        };
        if snapshot.promoter_change <= request.min_change_pct {
            continue;
        }
        let mut row = base.clone();
        row.name = Some(symbol);
        row.set_numeric("promoter_pct_latest", snapshot.promoter_pct_latest);
        row.set_numeric("promoter_pct_prev", snapshot.promoter_pct_prev);
        row.set_numeric("promoter_change", snapshot.promoter_change);
        if let Some(q) = snapshot.latest_quarter {
            row.set_text("latest_quarter", q);
        }
        if let Some(v) = snapshot.fii_pct_latest {
            row.set_numeric("fii_pct_latest", v);
        }
        if let Some(v) = snapshot.dii_pct_latest {
            row.set_numeric("dii_pct_latest", v);
        }
        out.push(row);
    }
    out.sort_by(|a, b| {
        b.numeric("promoter_change")
            .unwrap_or(f64::NEG_INFINITY)
            .partial_cmp(&a.numeric("promoter_change").unwrap_or(f64::NEG_INFINITY))
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    out.truncate(request.limit);
    Ok(out)
}

fn screen_us_insiders(
    universe: &[ScreenRow],
    request: &PromoterBuyRequest,
) -> anyhow::Result<Vec<ScreenRow>> {
    let Some(api_key) = std::env::var("FMP_API_KEY")
        .ok()
        .filter(|s| !s.trim().is_empty())
    else {
        anyhow::bail!(
            "Rust promoter-buys for US requires FMP_API_KEY; the Python yfinance insider-purchases fallback is not cloned in Rust"
        );
    };
    let client = Client::builder()
        .user_agent("Mozilla/5.0 (compatible; screener-rs/0.1)")
        .build()?;
    let cutoff = Utc::now().date_naive() - Duration::days(FMP_WINDOW_DAYS);
    if request.min_yf_net_pct.is_some() {
        anyhow::bail!(
            "--min-yf-net-pct depends on Yahoo insider-purchases percentages; Rust US promoter-buys currently uses FMP Form 4 share counts"
        );
    }
    let mut out = universe
        .par_iter()
        .filter_map(|base| {
            let symbol = base
                .text("name")
                .or_else(|| base.text("ticker"))
                .map(|s| bare_symbol(&s))?;
            let agg = fetch_fmp_aggregate(&client, &symbol, &api_key, cutoff).ok()??;
            if agg.net_shares_6m <= 0.0 {
                return None;
            }
            let mut row = base.clone();
            row.name = Some(symbol);
            row.set_numeric("fmp_net_shares_6m", agg.net_shares_6m);
            row.set_numeric("fmp_buy_shares_6m", agg.buy_shares_6m);
            row.set_numeric("fmp_sell_shares_6m", agg.sell_shares_6m);
            row.set_numeric("fmp_buy_trans_6m", agg.buy_trans_6m as f64);
            row.set_numeric("fmp_sell_trans_6m", agg.sell_trans_6m as f64);
            Some(row)
        })
        .collect::<Vec<_>>();
    out.sort_by(|a, b| {
        b.numeric("fmp_net_shares_6m")
            .unwrap_or(f64::NEG_INFINITY)
            .partial_cmp(&a.numeric("fmp_net_shares_6m").unwrap_or(f64::NEG_INFINITY))
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    out.truncate(request.limit);
    Ok(out)
}

fn fetch_fmp_aggregate(
    client: &Client,
    symbol: &str,
    api_key: &str,
    cutoff: NaiveDate,
) -> anyhow::Result<Option<FmpAggregate>> {
    let mut transactions = Vec::new();
    for page in 0..FMP_MAX_PAGES {
        let url = format!(
            "https://financialmodelingprep.com/api/v4/insider-trading?symbol={}&page={page}&apikey={}",
            urlencoding::encode(symbol),
            urlencoding::encode(api_key)
        );
        let rows: Value = client
            .get(url)
            .send()
            .with_context(|| format!("FMP insider request failed for {symbol}"))?
            .error_for_status()
            .with_context(|| format!("FMP insider request returned non-success for {symbol}"))?
            .json()
            .context("FMP insider JSON parse failed")?;
        let Some(page_rows) = rows.as_array() else {
            break;
        };
        if page_rows.is_empty() {
            break;
        }
        let oldest = page_rows.last().and_then(transaction_date);
        transactions.extend(page_rows.iter().cloned());
        if oldest.is_some_and(|day| day < cutoff) {
            break;
        }
    }
    Ok(aggregate_fmp_transactions(&transactions, cutoff))
}

fn aggregate_fmp_transactions(transactions: &[Value], cutoff: NaiveDate) -> Option<FmpAggregate> {
    let mut agg = FmpAggregate::default();
    for txn in transactions {
        let Some(day) = transaction_date(txn) else {
            continue;
        };
        if day < cutoff {
            continue;
        }
        let shares = txn
            .get("securitiesTransacted")
            .and_then(value_as_f64)
            .unwrap_or(0.0);
        let disposition = txn.get("acquistionOrDisposition").and_then(Value::as_str);
        let txn_type = txn
            .get("transactionType")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_uppercase();
        if txn_type.starts_with("P-") && disposition == Some("A") {
            agg.buy_shares_6m += shares;
            agg.buy_trans_6m += 1;
        } else if txn_type.starts_with("S-") && disposition == Some("D") {
            agg.sell_shares_6m += shares;
            agg.sell_trans_6m += 1;
        }
    }
    if agg.buy_trans_6m == 0 && agg.sell_trans_6m == 0 {
        None
    } else {
        agg.net_shares_6m = agg.buy_shares_6m - agg.sell_shares_6m;
        Some(agg)
    }
}

fn transaction_date(txn: &Value) -> Option<NaiveDate> {
    let raw = txn
        .get("transactionDate")
        .or_else(|| txn.get("filingDate"))
        .and_then(Value::as_str)?;
    NaiveDate::parse_from_str(raw.get(..10)?, "%Y-%m-%d").ok()
}

fn value_as_f64(value: &Value) -> Option<f64> {
    value
        .as_f64()
        .or_else(|| value.as_str().and_then(|s| s.replace(',', "").parse().ok()))
}

fn bare_symbol(raw: &str) -> String {
    raw.split_once(':')
        .map(|(_, rest)| rest)
        .unwrap_or(raw)
        .trim()
        .to_uppercase()
}

fn india_symbol(raw: &str) -> String {
    bare_symbol(raw)
        .trim_end_matches(".NS")
        .trim_end_matches(".BO")
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn aggregates_only_open_market_fmp_purchases_and_sales() {
        let cutoff = NaiveDate::from_ymd_opt(2026, 1, 1).unwrap();
        let rows = vec![
            serde_json::json!({"transactionDate": "2026-02-01", "securitiesTransacted": 100, "transactionType": "P-Purchase", "acquistionOrDisposition": "A"}),
            serde_json::json!({"transactionDate": "2026-02-02", "securitiesTransacted": 40, "transactionType": "S-Sale", "acquistionOrDisposition": "D"}),
            serde_json::json!({"transactionDate": "2026-02-03", "securitiesTransacted": 999, "transactionType": "A-Award", "acquistionOrDisposition": "A"}),
        ];
        let agg = aggregate_fmp_transactions(&rows, cutoff).unwrap();
        assert_eq!(agg.buy_shares_6m, 100.0);
        assert_eq!(agg.sell_shares_6m, 40.0);
        assert_eq!(agg.net_shares_6m, 60.0);
    }
}
